### PR TITLE
Cube build stage3 - JP-2096

### PR DESCRIPTION
**Pull request by jemorrison**
**Description**

Speeding up cube_build using numba.  Numba was installed using 'pip'
The purpose of this PR is to speed up cube_build using numba. Some of the modules in ifu_cube.py were moved out
of the class and made independent routines. Some of these routines were broken up into simpler routines. In this process the  weighting by the Miri psi option was removed. It is not being used and it was getting cumbersome to it as an option. Removing Miri ps weighting as an option also means that the resolution reference files is not longer needed. 

Including this PR in the JW pipeline requires numba to be added. But so far there seems little down side to including numba. It is fast, stable and easy to use.

### CODE REVIEW

**Comment by jemorrison** [comment]
PR ready for review.

**Comment by jemorrison** [comment]
See information on JP-2096 on numba and speed tests D Law has run with the code.

**Comment by jemorrison** [comment]
Once it is approved that numba can be used in JWST Pipeline I can make further speed improvements in the blotting routines and have them use numba. I am going to hold off on that until it is confirmed numba is fine to use the jwst pipeline

**Comment by hbushouse** [comment]
@jdavies-st or @eslavich any ideas what's killing the CI test "CI/Installed package with --pyargs"?

**Comment by jdavies-st** [comment]
If you search the log for "FAILURES"  you can see it is having trouble `asdf.open()` on a downloaded CRDS NIRISS WFSS reference file.

**Comment by jdavies-st** [comment]
We should be opening those reffiles in a cleaner way, i.e. with a `with` context manager so that if they fail to load, we don't get a cascade of errors.

**Comment by jemorrison** [comment]
@jdavies-st @nden if you have time now could you look this over or suggest someone else to look at it.
It is using numba to speed up cube_build

**Comment by jemorrison** [comment]
@hbushouse @nden @mcara 
This PR is ready for review. This PR pulled out the computational  python modules and converted them to c routines. There are now two c extensions 1. creating IFU cubes on sky 2. creating IFU cubes in detector plane.
I made a common set of c functions that both c extensions call in  cube_utilts.c 
As noted above in the comments - an unused weighting function - miripsf - was removed as an option. 
The unit tests run and the regression tests also run (with a few differences due to improvements I made when creating c routines).

**Comment by drlaw1558** [comment]
Following the usual means of checking out a PR, I'm getting errors about being unable to find the C modules.  E.g.,
from .cube_match_sky import cube_wrapper
returns
No module named jwst.cube_build.cube_match_sky

I'm probably just doing something wrong- do I need to do anything special to compile the code inside the src/ directory first or should python do that automatically?

**Comment by jemorrison** [comment]
To compile the c code. In the top jwst directory - the directory with setup.py
Type pip install -e .
That compiles the c code that is defined in setup.py
Let me you if you have problems with that

File: jwst/cube_build/src/cube_match_internal.c [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+  if (mem_alloc(ncube, &fluxv, &weightv, &varv, &ifluxv)) return 1;
+    
+  double set_zero=0.0;
+  // Set all data to zero
+  for (int i = 0; i < ncube; i++){
+    varv[i] = set_zero;
+    fluxv[i] = set_zero;
+    ifluxv[i] = set_zero;
+    weightv[i] = set_zero;
+  }
```
Comment (mcara) [inline jwst/cube_build/src/cube_match_internal.c:167]: I would suggest that 1) `mem_alloc` be renamed to something less generic and 2) use `calloc` in order to initialize arrays to 0. For example:
```suggestion
int alloc_flux_arrays(int nelem, double **fluxv, double **weightv, double **varv,  double **ifluxv) {
    const char *msg = "Couldn't allocate memory for output arrays.";
    
    // flux:
    if (!(*fluxv  = (double*)calloc(nelem, sizeof(double)))) {
        PyErr_SetString(PyExc_MemoryError, msg);
        goto failed_mem_alloc;
    }

    // weight:
    if (!(*weightv = (double*)calloc(nelem, sizeof(double)))) {
        PyErr_SetString(PyExc_MemoryError, msg);
        goto failed_mem_alloc;
    }

    // variance:
    if (!(*varv = (double*)calloc(nelem, sizeof(double)))) {
        PyErr_SetString(PyExc_MemoryError, msg);
        goto failed_mem_alloc;
    }
    // iflux
    if (!(*ifluxv = (double*)calloc(nelem, sizeof(double)))) {
        PyErr_SetString(PyExc_MemoryError, msg);
        goto failed_mem_alloc;
    }

    return 0;
    
failed_mem_alloc:
    free(*fluxv);
    free(*weightv);
    free(*varv);
    free(*ifluxv);
}

//_______________________________________________________________________
//  based on a 2-D drizzling type of algorithm find the contribution of
// each detector flux to the IFU cube
//_______________________________________________________________________

int match_detector_cube(int instrument, int naxis1, int naxis2, int nz, int npt, int ncube, int na,
			double crval_along, double cdelt_along, double crval3, double cdelt3,
			double *a1, double *a2,double *a3, double*a4,
			double *lam1, double *lam2, double*lam3, double *lam4,
			double *acoord, double *zcoord, int ss,
			double *pixel_flux, double *pixel_err,
			double **spaxel_flux, double **spaxel_weight, double **spaxel_var,double **spaxel_iflux){

  double *fluxv, *weightv, *varv;  // vectors for spaxel 
  double *ifluxv;  // vector for spaxel

  // allocate memory to hold output 
  if (alloc_flux_arrays(ncube, &fluxv, &weightv, &varv, &ifluxv)) return 1;
    
```

File: jwst/cube_build/src/cube_match_sky.c [hunk 2]
```diff
@@ -0,0 +1,1441 @@
...
+
+// routines used from cube_utils.c
+
+extern double sh_find_overlap(const double xcenter, const double ycenter, 
+		      const double xlength, const double ylength,
+			      double xPixelCorner[],double yPixelCorner[]);
+
+// Allocate the memory to for the spaxel arrays in the cube
+
+int mem_alloc(int nelem, double **fluxv, double **weightv, double **varv,  double **ifluxv) {
```
Comment (mcara) [inline jwst/cube_build/src/cube_match_sky.c:108]: Isn't this the same function as in `cube_match_internal.c`? Maybe it should be placed in a separate file and then included in both modules?

File: jwst/cube_build/src/cube_match_sky.c [hunk 3]
```diff
@@ -0,0 +1,1441 @@
...
+    int  *i;
+    const char *msg = "Couldn't allocate memory for output arrays.";
+
+    i = (int*)malloc(nelem * sizeof(int));
+    if (i) {
+        *idqv = i;
+    } else {
+        PyErr_SetString(PyExc_MemoryError, msg);
+        return 1;
+    }
```
Comment (mcara) [inline jwst/cube_build/src/cube_match_sky.c:164]:
```suggestion
    const char *msg = "Couldn't allocate memory for output arrays.";

    if (!(*idqv = (int*)calloc(nelem, sizeof(int)))) {
        PyErr_SetString(PyExc_MemoryError, msg);
        return 1;
    }
```

File: jwst/cube_build/src/cube_match_sky.c [hunk 4]
```diff
@@ -0,0 +1,1441 @@
...
+int dq_set_zero(int ncube, int **spaxel_dq){
+    int *idqv = NULL;  // int vector for spaxel
+
+    if (mem_alloc_dq(ncube, &idqv)) return 1;
+
+    // Set all data to zero
+    for (long i = 0; i < ncube; i++){
+      idqv[i] = 0;
+    }
+     *spaxel_dq = idqv;
```
Comment (mcara) [inline jwst/cube_build/src/cube_match_sky.c:633]: Assuming you have switched to `calloc` as suggested above
```suggestion
    if (mem_alloc_dq(ncube, spaxel_dq)) return 1;
```

File: jwst/cube_build/src/cube_match_sky.c [hunk 5]
```diff
@@ -0,0 +1,1441 @@
...
+
+  }
+  return status; 
+}
+
+//________________________________________________________________________________
+// Set the spaxel dq = 0. This is used when not determining the FOV on the sky for
+// setting the DQ plane. This is case for internalCal type cubes
+
+int dq_set_zero(int ncube, int **spaxel_dq){
```
Comment (mcara) [inline jwst/cube_build/src/cube_match_sky.c:624]: function name seems confusing/not informative.

File: jwst/cube_build/src/cube_match_sky.c [hunk 6]
```diff
@@ -0,0 +1,1441 @@
...
+	    int **spaxel_dq) {
+
+  int *idqv = NULL;  // int vector for spaxel
+
+    if (mem_alloc_dq(ncube, &idqv)) return 1;
+
+    // Set all data to zero
+    for (long i = 0; i < ncube; i++){
+      idqv[i] = 0;
+    }
```
Comment (mcara) [inline jwst/cube_build/src/cube_match_sky.c:657]:
```suggestion
    int *idqv;  // int vector for spaxel
    if (mem_alloc_dq(ncube, &idqv)) return 1;
```

File: jwst/cube_build/src/cube_match_sky.c [hunk 7]
```diff
@@ -0,0 +1,1441 @@
...
+  
+
+  int *idqv = NULL;  // int vector for spaxel
+
+  if (mem_alloc_dq(ncube, &idqv)) return 1;
+
+  // Set all data to zero
+  for (long i = 0; i < ncube; i++){
+    idqv[i] = 0;
+  }
```
Comment (mcara) [inline jwst/cube_build/src/cube_match_sky.c:755]:
```suggestion
  int *idqv;  // int vector for spaxel
  if (mem_alloc_dq(ncube, &idqv)) return 1;
```

File: jwst/cube_build/src/cube_match_sky.c [hunk 8]
```diff
@@ -0,0 +1,1441 @@
...
+  //     onto each of the IFU wavelength planes.
+
+  for (int w = 0; w  < nz; w++) {
+    int n_slice_found = 0;
+    for (int islice =1; islice< 31 ; islice++){
+      double c1_min;
+      double c2_min;
+      double c1_max;
+      double c2_max;
+      int status = 0;
```
Comment (mcara) [inline jwst/cube_build/src/cube_match_sky.c:767]:
```suggestion
      int status; 
```

File: jwst/cube_build/src/cube_match_sky.c [hunk 9]
```diff
@@ -0,0 +1,1441 @@
...
+    if (mem_alloc(ncube, &fluxv, &weightv, &varv, &ifluxv)) return 1;
+    
+    double set_zero=0.0;
+    // Set all data to zero
+    for (int i = 0; i < ncube; i++){
+      varv[i] = set_zero;
+      fluxv[i] = set_zero;
+      ifluxv[i] = set_zero;
+      weightv[i] = set_zero;
+    }
```
Comment (mcara) [inline jwst/cube_build/src/cube_match_sky.c:833]: This can be simplified as in `cube_match_internal.c` if using `alloc_flux_arrays`.

File: jwst/cube_build/src/cube_match_sky.c [hunk 10]
```diff
@@ -0,0 +1,1441 @@
...
```
Comment (mcara) [inline jwst/cube_build/src/cube_match_sky.c:1000]: Again this can be simplified as in `cube_match_internal.c` if using `alloc_flux_arrays`.

File: jwst/cube_build/src/cube_match_internal.c [hunk 11]
```diff
@@ -0,0 +1,484 @@
+/*
+The detector pixels are represented by a 'point could' on the sky. The IFU cube is
```
Comment (hbushouse) [inline jwst/cube_build/src/cube_match_internal.c:2]: typo: "could" -> "cloud"

File: jwst/cube_build/src/cube_match_internal.c [hunk 12]
```diff
@@ -0,0 +1,484 @@
...
+represented by a 3-D regular grid. This module finds the point cloud members contained
+in a region centered on the center of the cube spaxel. The size of the spaxel is spatial
```
Comment (hbushouse) [inline jwst/cube_build/src/cube_match_internal.c:4]: typo? Do you mean "size of the spaxel *in* spatial coords?

File: jwst/cube_build/src/cube_match_internal.c [hunk 13]
```diff
@@ -0,0 +1,484 @@
...
+coordinates is cdetl1 and cdelt2, while the wavelength size is zcdelt3.
```
Comment (hbushouse) [inline jwst/cube_build/src/cube_match_internal.c:5]: typo? should "zcdelt3" be just "cdelt3"?

Comment (hbushouse) [reply jwst/cube_build/src/cube_match_internal.c:5]: Oh, and "cdetl1" should be "cdelt1"

File: jwst/cube_build/src/cube_match_internal.c [hunk 14]
```diff
@@ -0,0 +1,484 @@
...
+This module uses the e modified shephard weighting method to determine how to  weight each point clold member
```
Comment (hbushouse) [inline jwst/cube_build/src/cube_match_internal.c:6]: do you really mean "the *e* modified shepard weighting" or is the "e" extraneous? And another instance of "clold" that should be "cloud".

File: jwst/cube_build/src/cube_match_sky.c [hunk 15]
```diff
@@ -0,0 +1,1441 @@
+/*
+The detector pixels are represented by a 'point could' on the sky. The IFU cube is
```
Comment (hbushouse) [inline jwst/cube_build/src/cube_match_sky.c:2]: "could" -> "cloud"

File: jwst/cube_build/src/cube_match_sky.c [hunk 16]
```diff
@@ -0,0 +1,1441 @@
...
+represented by a 3-D regular grid. This module finds the point cloud members contained
+in a region centered on the center of the cube spaxel. The size of the spaxel is spatial
+coordinates is cdetl1 and cdelt2, while the wavelength size is zcdelt3.
```
Comment (hbushouse) [inline jwst/cube_build/src/cube_match_sky.c:5]: All the same typos as the same paragraph in the previous module (cdetl1, zcdelt3, ...)

File: setup.cfg [hunk 17]
```diff
@@ -35,6 +35,7 @@ install_requires =
     gwcs>=0.16.1
     jsonschema>=3.0.2
     numpy>=1.17
+    numba>=0.50.0
```
Comment (hbushouse) [inline setup.cfg:38]: Now that we're using C extensions instead of numba, can this be removed?

File: jwst/cube_build/blot_cube.py [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+    flux_cube: median flux
+    blot_flux & blot_weight: blotted values of flux_cube to detector
+    """
+
+    roi_det = 1.0  # Just large enough that we don't get holes
+    xdistance = np.absolute(x_cube[ipt] - xcenter)
+    ydistance = np.absolute(y_cube[ipt] - ycenter)
+
+    index_x = np.where(xdistance <= roi_det)
+    index_y = np.where(ydistance <= roi_det)
```
[the remark this answers is not in the document]
Comment (jemorrison) [reply jwst/cube_build/blot_cube.py:33]: I am going to hold off making changes to blot_cube.py because I have another JP ticket to work just on blot cube after I get the c extensions in this PR committed. I will come back to these changes suggestions later this week.

File: jwst/cube_build/blot_cube_build.py [hunk 19]
```diff
@@ -338,30 +323,17 @@ def blot_images_nirspec(self):
...
-                        blot_weight[index2d] = blot_weight[index2d] + weight_distance
+                    blot_cube.blot_overlap(ipt, xstart,
+                                           xcenter, ycenter,
+                                           x_slice, y_slice,
+                                           flux_slice,
+                                           blot_xsize,
+                                           blot_flux, blot_weight)
                 ts1 = time.time()
-                log.debug(f"Time to map 1 slice  =  {ts1-ts0:.1f}")
+                log.debug(f"Time to blot 1 slice on NIRspec  =  {ts1-ts0:.1f}")
```
[the remark this answers is not in the document]
Comment (jemorrison) [reply jwst/cube_build/blot_cube_build.py:336]: it is for NIRSPEC. It can take several seconds per slice. Once we get it faster I will remove the debug timing

File: jwst/cube_build/cube_build_wcs_util.py [hunk 20]
```diff
@@ -111,27 +111,28 @@ def find_corners_MIRI(input, this_channel, instrument_info, coord_system):
...
+
+        if a1 < 0:
+            a1 = a1 + 360
+        if a1 > 360.0:
+            a1 = a1 - 360.0
+
+        if a2 < 0:
+            a2 = a2 + 360
+        if a2 > 360.0:
+            a2 = a2 - 360.0
```
[the remark this answers is not in the document]
Comment (jemorrison) [reply jwst/cube_build/cube_build_wcs_util.py:129]: nice! I did not know that.

**Comment by jemorrison** [comment]
@jdavies-st @eslavich 
The PR failed on CI/Installed package with --pyargs 
What does the mean  ? I looked at the details but I am a bit lost how to fix it

File: setup.cfg (hunk 17 above)
Comment (jemorrison) [reply setup.cfg]: yes I forgot I put that in

File: jwst/cube_build/src/cube_match_sky.c (hunk 16 above)
Comment (jemorrison) [reply jwst/cube_build/src/cube_match_sky.c:5]: fixed typos

File: jwst/cube_build/src/cube_match_sky.c (hunk 9 above)
Comment (jemorrison) [reply jwst/cube_build/src/cube_match_sky.c]: good suggestion. I moved alloc_flux_arrays to cube_utils.c

**Comment by jdavies-st** [comment]
It looks like it was getting a corrupted CRDS reference file.  I kicked off the CI again.  Hopefully it won't be corrupted this time.

File: jwst/cube_build/src/cube_match_internal.c (hunk 11 above)
Comment (jemorrison) [reply jwst/cube_build/src/cube_match_internal.c]: fixed

File: jwst/cube_build/src/cube_match_internal.c (hunk 12 above)
Comment (jemorrison) [reply jwst/cube_build/src/cube_match_internal.c]: fixed

File: jwst/cube_build/src/cube_match_internal.c (hunk 13 above)
Comment (jemorrison) [reply jwst/cube_build/src/cube_match_internal.c]: fixed

File: jwst/cube_build/src/cube_match_sky.c (hunk 10 above)
Comment (jemorrison) [reply jwst/cube_build/src/cube_match_sky.c]: done

File: jwst/cube_build/src/cube_match_sky.c (hunk 7 above)
Comment (jemorrison) [reply jwst/cube_build/src/cube_match_sky.c]: done

File: jwst/cube_build/src/cube_match_sky.c (hunk 6 above)
Comment (jemorrison) [reply jwst/cube_build/src/cube_match_sky.c]: done

File: jwst/cube_build/blot_cube_build.py [hunk 21]
```diff
@@ -338,30 +323,17 @@ def blot_images_nirspec(self):
...
-                        d2pix = np.array(y_slice[ipt] - ycenter[index_y])
-
-                        dxy = [(dx * dx + dy * dy) for dy in d2pix for dx in d1pix]
-                        dxy = np.sqrt(dxy)
-                        weight_distance = np.exp(-dxy)
-                        weighted_flux = weight_distance * flux_slice[ipt]
-                        index2d = [iy * blot_xsize + ix for iy in index_y[0] for ix in (index_x[0])]
-                        blot_flux[index2d] = blot_flux[index2d] + weighted_flux
-                        blot_weight[index2d] = blot_weight[index2d] + weight_distance
+                    blot_cube.blot_overlap(ipt, xstart,
```
[the remark this answers is not in the document]
Comment (jemorrison) [reply jwst/cube_build/blot_cube_build.py:329]: holding off on blot changes - I have opened a separate ticket on blotting

**Comment by jemorrison** [comment]
I made the changes suggested in the review. I reran the regression tests and  they look good. There are some expected differences because I improved  the DQ flags for edge cases and I made a few other little improvements. 
I think this is ready to be merged

**Comment by jdavies-st** [comment]
FYI the doc build is failing because the new C module cannot be imported for the doc build.  @eslavich do you recall what the issue was for C extensions and doc builds?

**Comment by hbushouse** [comment]
We made a slight change in docs/conf.py or something like that to remove a line that was causing it to look for things in a parent directory. Look at https://github.com/spacetelescope/jwst/pull/6207/files

**Comment by drlaw1558** [comment]
I'm seeing some larger changes in the data cubes than I would have expected given this just changes the language not the algorithm; let me investigate further and get back to you.  Did anything change that should have affected the total cube FOV?

**Comment by jemorrison** [comment]
@drlaw1558  I tweaked the DQ flagging.  Hopefully improving it near boundaries. I also tweaked- for MIRI - how the wavelength range of valid data is determined. There was a small bug in the old code.

**Comment by eslavich** [comment]
We merged a fix for the doc build failure in https://github.com/spacetelescope/jwst/pull/6230, so a rebase ought to get that unstuck.

**Comment by drlaw1558** [comment]
Ok, disregard my last comment- the changes that I was seeing were due to something changing earlier in the pipeline (that I'll have to run down elsewhere), not cube build.  Performance looks good to me; when the test is set up properly SCI results from running spec3 in multiple different modes look identical before/after this change but with vastly improved runtimes.

**Review by drlaw1558 — APPROVED** [review]
Focusing entirely on results, the performance of this PR looks good to me.  Running some test cases with dithered exposures in multiple bands through a variety of different kinds of cube building I get identical SCI results before/after this change but with vastly improved runtimes.  DQ arrays look improved.

**Comment by jemorrison** [comment]
@hbushouse can we merge this PR ?

**Review by hbushouse — APPROVED** [review]

File: jwst/cube_build/src/cube_match_internal.c [hunk 22]
```diff
@@ -0,0 +1,433 @@
...
+
+  double *fluxv, *weightv, *varv, *ifluxv ;  // vectors for spaxel 
+
+  // allocate memory to hold output 
+  if (alloc_flux_arrays(ncube, &fluxv, &weightv, &varv, &ifluxv)) return 1;
+    
+
+  // loop over each valid point on detector and find match to IFU cube based
+  // on along slice coordinate and wavelength
+  for (int ipixel= 0; ipixel< npt; ipixel++){
```
Comment (nden) [inline jwst/cube_build/src/cube_match_internal.c:120]: @jemorrison I believe you need to declare the variables in the beginning of the file as certain compilers won't like the definition within the for loop.

File: jwst/cube_build/src/cube_match_internal.c [hunk 23]
```diff
@@ -0,0 +1,433 @@
...
+    wave_corner[0] = lam1[ipixel];
+    wave_corner[1] = lam2[ipixel];
+    wave_corner[2] = lam3[ipixel];
+    wave_corner[3] = lam4[ipixel];
+
+    double along_min = 10000;
+    double wave_min = 10000;
+    double along_max = -10000;
+    double wave_max = -10000;
+    for (int j = 0; j< 4; j++){
```
Comment (nden) [inline jwst/cube_build/src/cube_match_internal.c:138]: Same comment as above

File: jwst/cube_build/src/cube_match_internal.c [hunk 24]
```diff
@@ -0,0 +1,433 @@
...
+    double MaxW = (wave_max - crval3) / cdelt3;
+    int iz1= (int) MinW;
+    int iz2 = round(MaxW);
+
+    if(iz1 < 0){ iz1 = 0;}
+    if (iz2 >= nz){iz2 = nz - 1;}
+
+    int nplane = naxis1 * naxis2;
+    // loop over possible overlapping cube pixels      
+    for(int zz =iz1; zz < iz2+1; zz++){
```
Comment (nden) [inline jwst/cube_build/src/cube_match_internal.c:165]: and here

File: jwst/cube_build/src/cube_match_internal.c [hunk 25]
```diff
@@ -0,0 +1,433 @@
...
+      double zcenter = zcoord[zz];
+      int istart = zz * nplane;
+      for (int aa= ia1; aa< ia2 + 1; aa++){
```
Comment (nden) [inline jwst/cube_build/src/cube_match_internal.c:168]: and here

**Comment by jemorrison** [comment]
@nden 
As I was moving all the definition to top of code. I see that I am I doing this:

nxy = nx * ny
int wave_slice_dq[nxy]

The code works fine doing this on my Mac. I am now thinking this may not be allowed - dynamically allocated the array. Should I change how I do this ?
